pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,292 @@
+package klinedriver
+
+import (
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	log "github.com/sirupsen/logrus"
+)
+
+type Bucket struct {
+	// StartTime is inclusive, EndTime is exclusive
+	// e.g. [StartTime, EndTime)
+	StartTime, EndTime types.Time
+	Interval           types.Interval
+	KLine              *types.KLine
+	Exposed            bool
+}
+
+// check if the currentTime is in [startTime, startTime + interval)
+func (bucket *Bucket) Contains(currentTime types.Time) bool {
+	if currentTime.Equal(bucket.StartTime.Time()) {
+		return true
+	}
+	return currentTime.After(bucket.StartTime.Time()) && currentTime.Before(bucket.EndTime.Time())
+}
+
+// FillingGapKLines returns the kline slices that are used to fill the gap between the bucket and the next bucket.
+// Note that it will also update the bucket's status including the start/end time and the kline.
+func (bucket *Bucket) FillingGapKLines(numShifts uint64) (kLines []*types.KLine) {
+	interval := bucket.Interval
+	if numShifts > 1 {
+		bucket.KLine.Open = bucket.KLine.Close
+		bucket.KLine.High = bucket.KLine.Close
+		bucket.KLine.Low = bucket.KLine.Close
+		bucket.KLine.Volume = fixedpoint.Zero
+		bucket.KLine.QuoteVolume = fixedpoint.Zero
+		bucket.KLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+		bucket.KLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		bucket.KLine.NumberOfTrades = 0
+	}
+	// fill the gap with empty trades
+	for i := uint64(1); i < numShifts; i++ {
+		bucket.StartTime = types.Time(bucket.StartTime.Time().Add(interval.Duration()))
+		bucket.EndTime = types.Time(bucket.EndTime.Time().Add(interval.Duration()))
+		bucket.KLine.StartTime = bucket.StartTime
+		bucket.KLine.EndTime = bucket.EndTime
+		// the opening kline
+		kline := *bucket.KLine // copy
+		kline.EndTime = types.Time(bucket.StartTime.Time().Add(time.Millisecond))
+		kline.Closed = false
+		kLines = append(kLines, &kline)
+		// the closing kline
+		kline = *bucket.KLine // copy again
+		kline.EndTime = bucket.EndTime
+		kline.Closed = true
+		kLines = append(kLines, &kline)
+	}
+	return
+}
+
+func (bucket *Bucket) accumulateTrade(trade *types.Trade) {
+	// expecting the trade time is in the same timezone as the kline
+	// if you are about to use the trade time, make sure to convert the timezone properly
+	// IMPORTANT: the trade should be treated as read-only here
+	tradeTimeInLocalTz := types.Time(trade.Time.Time().In(bucket.KLine.StartTime.Time().Location()))
```

Comment:
we don't specify time zone for the trade time from the implementation, they are all in UTC.